### PR TITLE
[FLINK-13376][datastream] ContinuousFileReaderOperator should respect…

### PR DESCRIPTION
**Pull request by ifndef-SleePy**
… semantics of BoundedOneInput

## What is the purpose of the change

* Currently `ContinuousFileReaderOperator` does not respect the semantics of `BoundedOneInput`
* If there are some operators chained with `ContinuousFileReaderOperator`, the `endInput` of these operators might be triggered before all datum finishing

## Brief change log

* Make `ContinuousFileReaderOperator` implementing `BoundedOneInput`
* Make `ContinuousFileReaderOperator` blocking in `endInput` instead of `close`
* Leave the blocking logic in `close` due to compatibility (I think the `BoundedOneInput.endInput` might be unstable)

## Verifying this change

* This change added an integration test

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 1]
```diff
@@ -196,6 +197,17 @@ public void dispose() throws Exception {
...
 		super.close();
 
+		waitReaderFinishing();
+
+		output.close();
+	}
+
+	@Override
+	public void endInput() throws Exception {
+		waitReaderFinishing();
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: I think the same fix is needed in `AsyncWaitOperator`

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java [hunk 2]
```diff
@@ -196,6 +197,17 @@ public void dispose() throws Exception {
...
+	}
+
+	private void waitReaderFinishing() throws InterruptedException {
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: nit: `waitReaderFinished()`? Or `waitSplitReaderFinished()`?

File: flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+		TestBoundedOneInputStreamOperator() {
+			// this operator must be chained with ContinuousFileReaderOperator
+			// that way, this end input would be triggered after ContinuousFileReaderOperator
+			chainingStrategy = ChainingStrategy.ALWAYS;
+		}
+
+		@Override
+		public void endInput() throws Exception {
+			isEnded = true;
```
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java]: nit: I would add also one more assertion here in the `endInput`, to check that expected number of elements was processed before  `endInput()` was invoked. (passing expected number of elements via the `TestBoundedOneInputStreamOperator` constructor).

File: flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		}
+
+		@Override
+		public void processElement(StreamRecord<String> element) throws Exception {
+			assertFalse(isEnded);
+			output.collect(element);
+		}
+	}
+
+	private static class NoOpSink implements SinkFunction<String> {}
```
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java]: There are something like 5 already existing `NoOpSink` classes. 

Could you extract them (a separate commit) to a common class, for example under `flink-streaming` module in test package `org.apache.flink.streaming.api.functions.sink.util`?

File: flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+		endInputChecking.addSink(new NoOpSink());
+
+		env.execute("ContinuousFileReaderOperatorITCase.testEndInput");
+	}
+
+	private static class TestBoundedOneInputStreamOperator extends AbstractStreamOperator<String>
+		implements OneInputStreamOperator<String, String>, BoundedOneInput {
+
+		private boolean isEnded = false;
```
Comment (pnowojski) [inline flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java]: nit: `hasEnded`?

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java (hunk 1 above)
Comment (ifndef-SleePy) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/source/ContinuousFileReaderOperator.java]: I don't like the feel that fixing similar issues one by one :(
Let's discuss under JIRA ticket.

File: flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java (hunk 3 above)
Comment (ifndef-SleePy) [reply flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java]: Make sense to me.

File: flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java (hunk 4 above)
Comment (ifndef-SleePy) [reply flink-tests/src/test/java/org/apache/flink/test/streaming/api/ContinuousFileReaderOperatorITCase.java]: Nice suggestion!
I just found there is already a `DiscardingSink`. So use this `DiscardingSink` instead of all `NoOpSink`s

**Comment by ifndef-SleePy** [comment]
Hi @pnowojski , thanks for reviewing. I have addressed the comments and updated the PR.

File: flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/DiscardingSink.java [hunk 6]
```diff
@@ -31,5 +31,5 @@
 	private static final long serialVersionUID = 1L;
 
 	@Override
-	public void invoke(T value) {}
+	public void invoke(T value, Context context) throws Exception {}
```
Comment (pnowojski) [inline flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/DiscardingSink.java]: Build failed because of this change (this is `@Public` class). Either let's keep it as it was before or you would have to somehow fix the api compatibility check (I don't remember at the moment how to handle it).

Comment (ifndef-SleePy) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/DiscardingSink.java]: Oops! I didn't expect it breaking the compatibility. I'll keep it as before.

Comment (pnowojski) [reply flink-streaming-java/src/main/java/org/apache/flink/streaming/api/functions/sink/DiscardingSink.java]: Neither did I ;)

**Comment by pnowojski** [comment]
Hey @ifndef-SleePy, did you change anything in the code in the last two recent force pushing? Or was it just rebasing?

**Comment by ifndef-SleePy** [comment]
Hi @pnowojski , 
1. I fixed a code style error which caused the Travis building failure. I squashed this fixing. So I forcedly update it.
2. However in the first force-updating, I got some unexpected testing codes involved in. And there is a python issue which caused the Travis building failure, https://issues.apache.org/jira/browse/FLINK-13488. So I revert these testing codes and did a rebasing. This is the second force-updating.
So in a nutshell, I fixed a code style error and rebased the master branch. But unfortunately, it seems that there is some wrong with the master branch.

**Comment by ifndef-SleePy** [comment]
One more question, what's your opinion of the way of addressing review comments since the new code style guide? Shall I do the squashing myself or just append some hotfix commits? Appending comments could keep the review history but leaves the squashing step to reviewer.

**Comment by ifndef-SleePy** [comment]
Rebase master again to solve the python Travis building failure issue

**Comment by pnowojski** [comment]
Thanks for the explanation @ifndef-SleePy. I was asking because I noticed the build failures and I was wandering whether I've missed something during the reviews.

> One more question, what's your opinion of the way of addressing review comments since the new code style guide? Shall I do the squashing myself or just append some hotfix commits? Appending comments could keep the review history but leaves the squashing step to reviewer.

There is no clear policy in this regard. I personally, as a contributor when I submit my PRs, I usually squash the fixup commits immediately, so usually I do not publish fixup commits in the PR. Where **usually** means if the fixup is trivial, didn't touch new code, etc.

Sometimes I publish fixup commit (that should be squashed before merging) if it contained something not trivial. I know that others contributors/committers have different practises in this regard (some of them are always publishing fixup commits and squash them only at the end).

As a reviewer I very rarely manually checkout the PR, so I usually ask the contributor to squash the appropriate commits together before I merge something (again, exception to this rule if the end state of the PR should be just a single commit, in that case I can squash them from the github ui with a click of one button, so I do not have to ask contributor to that ;) ).

Can you let me know once the travis is green?

**Comment by ifndef-SleePy** [comment]
Hi @pnowojski , thank you very much for feedback. Got your point. I was afraid my force-updating might mislead you.
WRT travis building, I don't know why it's so unlucky. It failed again due to unstable `KafkaITCase`. I have left messages under relevant jira tickets. I wish it could pass this time.

**Comment by pnowojski** [comment]
@ifndef-SleePy can you link the jira tickets for the last two failures?

**Comment by ifndef-SleePy** [comment]
@pnowojski , of course.
1. https://issues.apache.org/jira/browse/FLINK-12033
2. https://issues.apache.org/jira/browse/FLINK-12396

The python relevant error should caused by these issues below, however it's a bit messy.
1. https://issues.apache.org/jira/browse/FLINK-13488
2. https://github.com/apache/flink/commit/6dac7a7f7fd72aba903ac339f9f6bfcd44897d39, I believe

**Comment by ifndef-SleePy** [comment]
Hi @pnowojski , it failed again...
This time, the cause is https://issues.apache.org/jira/browse/FLINK-9900 which is also followed by me. I will talk to Till to merge that PR. It happens too frequently.

**Comment by pnowojski** [comment]
@ifndef-SleePy ok, I will merge this PR as those failures are almost definitely unrelated and across couple of those builds all tests had a chance to pass at least once.

Could you remove the two empty commits from the branch?

**Comment by ifndef-SleePy** [comment]
@pnowojski , done

**Comment by pnowojski** [comment]
Thanks! merging. 

Indeed those test failures are problematic :/

**Comment by ifndef-SleePy** [comment]
Haha, it bothers me a lot. I think these cases would be resolved soon. 
Thanks for reviewing!
